### ISSUE TITLE
FR: Action.ReturnType should look for ResponseType attribute when return type IHttpActionResult

### ISSUE THREAD

**Issue by gregveres** [NeVeSpl/NTypewriter#15]
As I mentioned in a previous request, I am still on ASP.Net MVC and haven't moved to .Net core yet.  So my api controller actions all look like this:

```
        [IsAjaxRequest]
        [HttpPut]
        [Route("{adId}/image")]
        [ResponseType(typeof(AdImageViewModel.AdImage))]
        public async Task<IHttpActionResult> ReplaceImage(int adId)
        {
            ...
        }
```

The real type of my method is encoded in the ResponseType attribute. 

In the .Net Core world, this is done like this [article ](https://docs.microsoft.com/en-us/aspnet/core/web-api/action-return-types?view=aspnetcore-5.0) points out. The attribute is different and the return type is different, but the concept is the same.

It would be great if Action.ReturnType handled this for ASP.Net MVC Web Api 2.0 and for .Net Core.

**Comment by gregveres** [comment]
@NeVeSpl I am trying to tackle this problem. But to do it right, I need some unit tests that test against an MVC Web Api 2.0 project. 

I have tried to add such a project to the solution as project: "Tests.Assets.WebApiMVC20". Then I created a new unit test under the project: NTypewriter.CodeModel.Functions.Tests/Action called ActionFunctionsTests_WebApi20.cs.

Then I changed the project path in the Initialize() function:

```csharp
        [TestInitialize]
        public async Task Initialize()
        {
            var config = new CodeModelConfiguration().FilterByNamespace("Tests.Assets.WebApiMVC20");
            data = await CreateCodeModelFromProject(config, "Tests.Assets.WebApiMVC20");
            settings = new Configuration();
        }
```

This seems like it works. I debugged through it and it is picking up the csproj file that is there and none of the functions claim there are errors, but the code model that gets returned has zero classes in it, so it is clearly not picking put the symbols. 

Do you have any suggestions on how to get the unit test to pick up the code symbols from that project? Is there something else I need to change in the setup? 

I assume you are sleeping now so I will keep plugging away.

**Comment by gregveres** [comment]
I am giving up for now. I can't figure out how to get the symbols for the MVC Asp.Net Web Api 2.0 project loaded. If I could, then I already have the unit tests in place to test the desired behaviour and I would be able to make the necessary changes. 

But in order to make the changes, I have to do some debugging to see what values I am going to get out of the attribute to use as the return type.

**Comment by gregveres** [comment]
I figured out how to get the unit tests running on the MVC project. It turns out that the unit test project is running under a .Net Core env but the environment variables in the MVC csproj file were expecting the MVC .net env and so things weren't where they were supposed to be. 

I was able to remove a reference to the webtargets project that gets pulled in. This was the one causing the failure. Without it being referenced, I was able to get the project loaded and see the symbols. I now have to figure out the right changes to get the tests passing, but I didn't want you spending any time looking at it now that I am making progress again.

**Comment by gregveres** [comment]
Hmm, now that I see what I am dealing with, I don't really see how I am going to get the data out of the attribute. 
I have a generic Attribute. I could cast it to ResponseTypeAttribute, but we aren't linking to the WebApi2.0 dll that contains the code for it. 

Are you using reflection elsewhere in the code? That's the only way I can see getting the value out. Looking at the attribute in the debugger, I can see that there is a public get method called ResponseType that if I could get to it and call it, hopefully it would return the value. 

Do you have any thoughts on how to approach this? I am calling it quits for tonight and I will look at it again tomorrow. 

I am also seeing a lot of errors in my unit tests for all the other Action methods that I will look into tomorrow. I created a WebApi 2.0 controller with 8 methods to try and get a bunch of different scenerios, such as:
- different http method types (these all work as expected)
- different route attributes and route prefix combinations (the urls don't seem to be working as I expect)
- different combinations of args that come from the route, the url and the body (these don't seem to work as I expect either).

I need to dig in to see if it is my test that is wrong, my controller that is wrong or the code that is wrong.

**Comment by gregveres** [comment]
I decided to look at the Action.Url code to see how you extract a value from RoutePrefix. I see that the Arguments array should be set, but in my debug session the Arguments is null. 

I am going to code the ReturnType using the arguments to see if I can extract it that way and then run it outside of the test harness to see if that works and the problem is really the test harness.

**Comment by NeVeSpl** [comment]
Reflection is not an option, you do not have access to it from Roslyn API. 

The test harness is set up to work with net5.0/net core, as you already know it won't work with older ones. Sure, you can try to change that, but I remember how painful it wast to make it work, so it is fine by my if there aren't any test for this issue.

**Comment by NeVeSpl** [comment]
Implementation is ready, and it is waiting for your tests. If you will not succeed with making tests working for MVC Asp.Net Web Api 2.0 project consider adding only test against [ProducesResponseType] attribute from .net core.

**Comment by gregveres** [comment]
I am trying a different approach but I have run into a null pointer crash in MSBuild stuff. I might commit it to my fork and then post an issue on the roslyn github. I see another issue there that is similar but it didn't look like the person provided a reproducible use case. 

I will take a crack at the unit test for [ProducesResponseType] then.

**Comment by gregveres** [comment]
I have tried your change and it mostly works. I am going to create some unit tests based on ProducesResponseType to see if I can reproduce the case where it doesn't work. It could also be a problem with .nt script. 

What I am seeing is that [ResponseType(typeof(List<someType>))] is showing up as void. But, like I said, it might be a problem in my script.

**Comment by gregveres** [comment]
I think your change works perfectly. I think the problem is in my script. 

The first unit tests I did nicely showed an IEnumerable<someType>, so I think  this is working great. 
At this point, I would go ahead and release a new version publicly. I think I am able to completely generate all my types, enums and services with the code that is currently checked in. Oh, I mean if you accept the outstanding PR for the withoutBodyParameter change.

I have to step out, but when I come back, I will do a good set of unit tests for this change and submit it as a PR.

**Comment by gregveres** [comment]
OK, I was an idiot. I thought that I had changed my script to use your new code, but I had not.

I have now and the ReturnType is always coming back as null. I think this lines up with my unit testing for this feature. For some reason, when the attributes are attached to the method in the ReturnType call, the attributes are incomplete and do not include the arguments. This is true during unit testing and it seems to be true at run time as well. 

Here is the section of script I am using:

```
export class {{serviceName}} {
  {{for method in controller.Methods | Symbols.ThatArePublic
		methodName = method.BareName | String.ToLowerFirst 
		routeName =  methodName | String.Append "Route"
		debugInfo = methodName | String.Append " - "
		returnType = (method | Action.ReturnType) ?? "void"

		url = method | Action.Url
		bodyParameterName = (method | Action.BodyParameter)?.Name ?? "null"
		parameters = method | Action.Parameters | Parameters.ToTypeScript | Array.Join ", "
		urlParams = method | Action.Parameters true
		routeFnParams = urlParams | Parameters.ToTypeScript | Array.Join ", "
		routeCallParmas = urlParams | Array.Map "Name" | Array.Join ", "
  }}

  public static {{routeName}} = ({{routeFnParams}}) => `{{url}}`;
  public static {{methodName}}({{parameters}}) : Promise<{{returnType}}> {
    return MakeRequest<{{returnType}}>("{{method | Action.HttpMethod}}", this.{{routeName}}({{routeCallParmas}}), {{bodyParameterName}});
  }
  {{end}}
}
```

and this is the output from it. Note the void return, which indicates a null was returned from ReturnType.

```
 public static initialAppInfoRoute = () => `api/Member/InitialAppInfo`;
  public static initialAppInfo() : Promise<void> {
    return MakeRequest<void>("get", this.initialAppInfoRoute(), null);
  }
```
The return should have been InitialAppInfo

What I was using before your ReturnType name change is:

```
		returnType = (method | MethodReturnType) ?? "void"
```

and MethodReturnType was a function I defined like this:

```
{{- func MethodReturnType(method)
		# check for the use of the ResponseType attribute
		for attr in method.Attributes
			if (attr.Name == "ResponseType") 
				ret ResponseTypeValue attr
			end
		end
		ret null
    end
}}
{{- func ResponseTypeValue(attr)
		typeName = attr.Arguments[0].Value
		for class in data.Classes
		  if (class.FullName == typeName)
		    ret class
	      end
		end
		ret null
	end
}}
```

This actually worked. It spit out InitialAppInfo as the return type. So the attribute does have arguments and I checked, the argument is called "responseType". 

I have more thoughts on this that I will put down here after some more experimentation.

**Comment by gregveres** [comment]
Ok, I have just finished updating my script to properly generate the return type - without using the Action.ReturnType function call. That is because I know that all of my api action methods use the ResponseType attribute, so I know that I can safely extract the return type from that. 

You might already have your head wrapped around the attribute, but it took me a while to get there. 

The arguments of the attributes are just strings. In the .nt script they are strings with no way to convert them to Types. I see in your code you are converting the argument into an IType. Is that valid? I can't tell through debugging the unit test because the unit test always returns zero arguments. I don't know if that is a problem of the unit test or if that is always the case, even at run time. 

But the way that I have tackled the problem in the .nt script is that I search for ResponseType attribute of the method, then I take the Value of that attribute's first argument and I treat it as a type name. However, since I use a lot of List<> types as return types, I have to deal with that. And since I can't create a new Type that is List, with another type inside it (as an enumerable, generic), I had to settle for stripping the list, then searching the list of data.Classes for a match on the inner type. What I return is an object that includes the inner class and a flag for isArray. 

From there I can then create my import of the inner class and I can create my return type as innerClass[], or innerClass, depending on the value of my isArray flag. 

So, this all works for me and I think I am happy with NTypewriter as is (if you release a version with the Parameter withoutBodyParameter.)

As for this situation, I don't know how to proceed. I kind of feel that the unit test is correct and that the attributes that the method has attached don't actually include any of the arguments. So maybe there is a problem in the parsing code or building up the model. I don't think the problem is in the Action.ReturnType code. 

I am done for the night, but I might take another look tomorrow. If ReturnType was actually able to return a proper type for the case of [ResponseType(ScalarValue)] and return a proper List type with isEnumerable set and ArrayType set for [ResponseType(List<ScalarValue)], then it would definitely shorten my .nt script and I assume make it run faster because things would be executing in C# more than in Scriban.
